soot-infoflow-collections/src/soot/jimple/infoflow/collections/codeOptimization/StringResourcesResolver.java

```diff
@@ -0,0 +1,133 @@
+    }
+
+    @Override
+    public void run(InfoflowManager manager, Collection<SootMethod> excluded, ISourceSinkManager sourcesSinks,
+                    ITaintPropagationWrapper taintWrapper) {
+        // This optimization enables more constant propagation. It doesn't make sense to perform it without
+        // constant propagation enabled.
+        if (manager.getConfig().getCodeEliminationMode() == InfoflowConfiguration.CodeEliminationMode.NoCodeElimination)
+            return;
+
+        // Only android apps contain resources
+        if (fileName == null)
+            return;
+
+        long beforeOptimization = System.nanoTime();
+
+        ARSCFileParser parser;
+        try {
+            parser = ARSCFileParser.getInstance(new File(fileName));
+        } catch (IOException e) {
+            logger.error("Could not parse the ARSC file! Aborting string resource resolving...", e);
+            return;
+        }
+        assert parser != null;
+
+        SootClass contextClass = Scene.v().getSootClassUnsafe(CONTEXT_CLASS);
+        if (contextClass == null) {
+            logger.error("Could not load class " + CONTEXT_CLASS + ". Aborting string resource resolving...");
+            return;
+        }
+
+        // First pass: Collect all statements to be replaced
+        ReplacementCandidates rcs = new ReplacementCandidates();
+        for (QueueReader<MethodOrMethodContext> rdr = Scene.v().getReachableMethods().listener(); rdr.hasNext();) {
+            MethodOrMethodContext sm = rdr.next();
+            SootMethod method = sm.method();
+            if (method == null || !method.isConcrete() || excluded.contains(method))
+                continue;
+
+            UnitPatchingChain chain = method.retrieveActiveBody().getUnits();
+            for (Unit unit : chain) {
+                Stmt stmt = (Stmt) unit;
+                // We only care for calls to a method where the returned value is used
+                if (!stmt.containsInvokeExpr() || !(stmt instanceof AssignStmt))
+                    continue;
+
+                // Check whether the invoke expressions calls a method we care about
+                SootMethod callee = stmt.getInvokeExpr().getMethod();
+                String subSig = callee.getSubSignature();
+                if (!manager.getHierarchy().isSubclass(callee.getDeclaringClass(), contextClass)
+                        || !subSig.equals(GET_STRING_SUBSIG))
+                    continue;
+
+                // Extract the resource id
+                Value arg0 = stmt.getInvokeExpr().getArg(0);
+                // We expect an integer constant here
+                if (!(arg0 instanceof IntConstant))
+                    continue;
+                int resourceId = ((IntConstant) arg0).value;
+
+                // Get the string for the given resource id
+                ARSCFileParser.AbstractResource res = parser.findResource(resourceId);
+                if (!(res instanceof ARSCFileParser.StringResource))
```

Comment:
Gibt es einen Grund, hier nur Stings zu supporten? Es gibt z.B. auch Integer-Konstanten in der ASRC-Datenbank.